conjure-error/src/error.rs

```diff
@@ -0,0 +1,380 @@
+use conjure_object::Value;
+use parking_lot::Mutex;
+use serde::Serialize;
+use std::borrow::Cow;
+use std::collections::hash_map::{self, HashMap};
+use std::error;
+use std::fmt;
+use std::ops::Index;
+use std::time::Duration;
+
+use crate::{ErrorType, Internal, SerializableError};
+
+/// Information about a throttle error.
+#[derive(Debug)]
+pub struct ThrottleError {
+    duration: Option<Duration>,
+}
+
+impl ThrottleError {
+    /// Returns the amount of time the client should wait before retrying, if provided.
+    #[inline]
+    pub fn duration(&self) -> Option<Duration> {
+        self.duration
+    }
+}
+
+/// Information about an unavailable error.
+#[derive(Debug)]
+pub struct UnavailableError(());
+
+/// Information about the specific type of an `Error`.
+#[derive(Debug)]
+pub enum ErrorKind {
+    /// A general service error.
+    Service(SerializableError),
+    /// A QoS error indicating that the client should throttle itself.
+    Throttle(ThrottleError),
+    /// A QoS error indicating that the server was unable to handle the request.
+    Unavailable(UnavailableError),
+    #[doc(hidden)]
+    __NonExhaustive,
+}
+
+#[derive(Debug)]
+struct Inner {
+    cause: Box<dyn error::Error + Sync + Send>,
+    cause_safe: bool,
+    kind: ErrorKind,
+    safe_params: HashMap<Cow<'static, str>, Value>,
+    unsafe_params: HashMap<Cow<'static, str>, Value>,
+    backtraces: Vec<Backtrace>,
+}
+
+/// A standard error type for network services.
+///
+/// An error consists of several components:
+///
+/// * The cause of the error, represented as a type implementing the Rust `Error` trait. The cause can either be
+///     declared safe or unsafe to log.
+/// * The error's kind, indicating how the service should handle the error e.g. in a response to a client.
+/// * Backtraces, including one taken at the time the error was created.
+/// * Parameters adding extra context about the error. They can be declared either safe or unsafe to log.
+///
+/// Note that this type does *not* implement the standard library's `Error` trait.
```

Comment:
Why doesn't this `Error` implement the standard library's `Error` trait?

Reply:
It's not intended to be used as one - once you starting using `Error` we generally expect the remainder of the call chain to deal with that type specifically.